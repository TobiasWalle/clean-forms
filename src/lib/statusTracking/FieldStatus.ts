export interface FieldStatusArguments {
  dirty: boolean;
  touched: boolean;
}

export class FieldStatus {
  public readonly dirty: boolean;
  public readonly pristine: boolean;
  public readonly touched: boolean;
  public readonly untouched: boolean;

  constructor(args: FieldStatusArguments) {
    const { dirty, touched } = args;

    this.dirty = dirty;
    this.pristine = !dirty;
    this.touched = touched;
    this.untouched = !touched;
  }
}

export const DEFAULT_FIELD_STATUS: FieldStatus = new FieldStatus({ dirty: false, touched: false });

export function cloneFieldStatus<Children = any>(statusToClone: FieldStatus, overrideArgs: Partial<FieldStatusArguments>): FieldStatus {
  return new FieldStatus({
    ...(statusToClone as any),
    ...overrideArgs
  });
}
