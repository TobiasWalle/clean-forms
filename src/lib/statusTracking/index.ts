export * from './FieldStatus';
export * from './FieldStatusMapping';
