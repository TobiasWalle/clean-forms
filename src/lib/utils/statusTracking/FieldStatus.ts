export interface FieldStatus {
  valid: boolean;
  inValid: boolean;
  pristine: boolean;
  dirty: boolean;
  untouched: boolean;
  touched: boolean;
  error?: string;
}

export const DEFAULT_FIELD_STATUS: FieldStatus = {
  valid: true,
  inValid: false,
  pristine: true,
  dirty: false,
  untouched: true,
  touched: false
};
