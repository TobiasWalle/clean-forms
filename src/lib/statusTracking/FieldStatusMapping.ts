import { FieldStatus } from './FieldStatus';

export interface FieldStatusMapping {
  [path: string]: FieldStatus;
}
