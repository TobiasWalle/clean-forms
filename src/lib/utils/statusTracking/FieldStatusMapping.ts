import { FieldStatus } from './FieldStatus';

export type FieldStatusMapping<Model> = {
  [key in keyof Model]?: FieldStatus | FieldStatusMapping<Model>;
};
