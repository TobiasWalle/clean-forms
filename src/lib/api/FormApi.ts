import { selectDeep } from '../utils';
import { DEFAULT_FIELD_STATUS, FieldStatus } from '../utils/statusTracking/FieldStatus';
import { FieldStatusMapping } from '../utils/statusTracking/FieldStatusMapping';
import { ValidationDefinition, FieldError, FieldErrorMapping } from '../utils/validation';

export interface FormState<Model> {
  model: Model;
  status?: FieldStatusMapping;
}

export class FormApi<Model> {
  public get status(): FieldStatusMapping {
    return this.state.status || {};
  }

  public get model(): Model {
    return this.state.model;
  }

  public constructor(
    private readonly state: FormState<Model>,
    public readonly validationDefinition: ValidationDefinition<Model> = {},
    public readonly fieldErrorMapping: FieldErrorMapping = {}
  ) {
  }

  public getFieldValue(path: string): any {
    return selectDeep({ object: this.model, path});
  }

  public getFieldStatus(fieldId: string): FieldStatus {
    return this.status[fieldId] || DEFAULT_FIELD_STATUS;
  }
}
