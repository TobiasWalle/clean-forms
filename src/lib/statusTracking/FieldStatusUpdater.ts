import { FieldRegister, Path } from '../utils/FieldRegister';
import { cloneFieldStatus, DEFAULT_FIELD_STATUS, FieldStatus, FieldStatusArguments } from './FieldStatus';
import { FieldStatusMapping } from './FieldStatusMapping';

export class FieldStatusUpdater {
  constructor(
    private fieldRegister: FieldRegister
  ) {
  }

  public markAsDirty(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    return this.updateField(status, path, { dirty: true });
  }

  public markAsTouched(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    return this.updateField(status, path, { touched: true });
  }

  public markAllAsTouched(status: FieldStatusMapping): FieldStatusMapping {
    return Object.keys(status)
      .reduce((result, path) => {
        return this.markAsTouched(result, path);
      }, status);
  }

  private updateField(
    status: FieldStatusMapping,
    path: Path,
    update: Partial<FieldStatusArguments>
  ): FieldStatusMapping {
    if (!this.fieldRegister.includesPath(path)) {
      return status;
    }
    const oldStatus = status[path];
    if (!oldStatus) {
      throw new ReferenceError(`Couldn't find path "${path}" in status mapping ${JSON.stringify(status)}`);
    }
    return { ...status, [path]: cloneFieldStatus(oldStatus, update) };
  }

  public addIfFieldNotExists(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    if (status[path] == null) {
      return {
        ...status,
        [path]: DEFAULT_FIELD_STATUS
      };
    } else {
      return status;
    }
  }

  public removeIfFieldExists(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    const copy = { ...status };
    delete copy[path];
    return copy;
  }
}
