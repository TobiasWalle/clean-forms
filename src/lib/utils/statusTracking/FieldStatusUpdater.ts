import { FieldRegister, Path } from '../FieldRegister';
import { cloneFieldStatus, DEFAULT_FIELD_STATUS, FieldStatusArguments } from './FieldStatus';
import { FieldStatusMapping } from './FieldStatusMapping';

export class FieldStatusUpdater {
  constructor(
    private fieldRegister: FieldRegister
  ) {
  }

  public markAsDirty(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    return this.updateField(status, path, {dirty: true});
  }

  public markAsTouched(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    return this.updateField(status, path, {touched: true});
  }

  private updateField(
    status: FieldStatusMapping,
    path: Path,
    update: Partial<FieldStatusArguments>
  ): FieldStatusMapping {
    if (!this.fieldRegister.includesPath(path)) {
      return status;
    }
    const oldStatus = status[path];
    if (!oldStatus) {
      throw new ReferenceError(`Couldn't find path "${path}" in status mapping ${JSON.stringify(status)}`);
    }
    return { ...status, [path]: cloneFieldStatus(oldStatus, update) };
  }

  public addIfFieldNotExists(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    if (status[path] == null) {
      return {
        ...status,
        [path]: DEFAULT_FIELD_STATUS
      };
    } else {
      return status;
    }
  }

  public removeIfFieldExists(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    const copy = { ...status };
    delete copy[path];
    return copy;
  }
}
