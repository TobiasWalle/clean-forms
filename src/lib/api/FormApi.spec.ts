import { DEFAULT_FIELD_STATUS, FieldStatus } from '../utils/statusTracking/FieldStatus';
import { FieldStatusMapping } from '../utils/statusTracking/FieldStatusMapping';
import { FormApi } from './FormApi';

describe('FormApi', () => {
  let model: any;
  let status: FieldStatusMapping;
  let api: FormApi<any>;
  let aFieldStatus: FieldStatus;

  beforeEach(() => {
    model = { a: 1, b: 2 };
    aFieldStatus = new FieldStatus({ dirty: true, touched: true });
    status = { a: aFieldStatus };
    api = new FormApi<any>(
      { model, status },
      {},
      { a: 'Error' }
    );
  });

  it('should provide status', () => {
    expect(api.status).toBe(status);
  });

  it('should provide model', () => {
    expect(api.model).toBe(model);
  });

  it('should get value for a specific field', () => {
    expect(api.getFieldValue('a')).toBe(1);
    expect(api.getFieldValue('b')).toBe(2);
    expect(() => api.getFieldValue('c')).toThrowErrorMatchingSnapshot();
  });

  it('should get status for a specific field', () => {
    expect(api.getFieldStatus('a')).toBe(aFieldStatus);
    expect(api.getFieldStatus('b')).toBe(DEFAULT_FIELD_STATUS);
    expect(api.getFieldStatus('c')).toBe(DEFAULT_FIELD_STATUS);
  });

  it('should get error for a specific field', () => {
    expect(api.getFieldError('a')).toBe('Error');
    expect(api.getFieldError('b')).toBe(undefined);
  });
});
