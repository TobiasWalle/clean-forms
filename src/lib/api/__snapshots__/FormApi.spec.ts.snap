// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormApi should get value for a specific field 1`] = `"The key \\"c\\" does not exits on item {\\"a\\":1,\\"b\\":2}."`;
